### PR TITLE
Setup CI/CD

### PR DESCRIPTION
**Pull request by gutakk**
## What happened
Setup CI/CD to make sure that this app will pass all tests and then continuous deploy to staging and prod server.
- Setup CI using Github Actions (followed Semaphore pipeline).
- Using Github Container Registry as Docker Registry.
- Setup Heroku app (both Staging and Production).
- Setup CD (followed Semaphore pipeline and run only when push to master or development branch).
- Deployed Rails app to staging Heroku server.
- Setup simple rails signup page to make sure that CD deployed correctly.
 

## Insight
`N/A`

 
## Proof Of Work
- CI passed (build, pull, unit test and system test)
Screen Shot 2563-09-17 at 15 01 17
Screen Shot 2563-09-17 at 15 01 27
- CD passed
Screen Shot 2563-09-17 at 15 02 32
- Able to open simple [signup page](https://ruby-google-scraper-staging.herokuapp.com/signup) on Heroku staging

### CODE REVIEW

File: .github/workflows/cd.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+name: Deploy
```
Comment (hoangmirs) [inline .github/workflows/cd.yml]: I think we can name this workflow as `deploy` instead of `cd`

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,58 @@
+name: CI
```
Comment (hoangmirs) [inline .github/workflows/ci.yml]: I think we can name this workflow as `test` instead of `ci`

File: .github/workflows/cd.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+          docker-compose build
+          docker-compose push web
+
+      - name: Heroku login
+        run: heroku container:login
+  
+      - name: Heroku release
+        run: |
+          heroku container:push --arg DOCKER_REGISTRY=$DOCKER_REGISTRY,DOCKER_IMAGE=$DOCKER_IMAGE,BRANCH_TAG=$BRANCH_TAG --recursive
+          heroku container:release web worker
```
Comment (hoangmirs) [inline .github/workflows/cd.yml]: Could you add a new line here?

File: .github/workflows/cd.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  HEROKU_API_KEY: ${{ secrets.HEROKU_API_KEY }}
+  HEROKU_APP: ruby-google-scraper-staging
+
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: nimblehq/branch-tag-action@v1
```
Comment (hoangmirs) [inline .github/workflows/cd.yml]: This action released a new version, you can use `v1.1` instead.
You can check the changes on this link:
https://github.com/marketplace/actions/branch-tag-action

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        run: docker-compose pull
+      
+      - name: Linters
+        run: docker-compose run test yarn lint
+
+      - name: Unit tests
+        run: docker-compose run test bundle exec rspec --exclude-pattern "spec/systems/**/*_spec.rb" --profile
+
+      - name: System tests
+        run: docker-compose run test bundle exec rspec spec/systems --profile
```
Comment (hoangmirs) [inline .github/workflows/ci.yml]: Could you add a new line here?

File: Dockerfile [hunk 6]
```diff
@@ -15,7 +15,7 @@ ARG MAILER_DEFAULT_PORT=3000
 ENV ASSET_HOST=$ASSET_HOST \
     BUILD_ENV=$BUILD_ENV \
     BUNDLE_JOBS=4 \
-    BUNDLE_PATH="/bundle" \
+    BUNDLE_PATH="/bundle"\
```
Comment (hoangmirs) [inline Dockerfile]: Why did you remove the space?

File: app/controllers/signup_controller.rb [hunk 7]
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
```
Comment (hoangmirs) [inline app/controllers/signup_controller.rb]: I think we don't need the signup on this PR, you can remove it after testing

Comment (hoangmirs) [reply app/controllers/signup_controller.rb]: and for the signup, I think it should be `UsersController` as in this controller, we want to create a new user

File: Dockerfile (hunk 6 above)
Comment (gutakk) [reply Dockerfile]: Removed by mistake. Solved

File: app/controllers/signup_controller.rb (hunk 7 above)
Comment (gutakk) [reply app/controllers/signup_controller.rb]: Make sense, since signup is not a part of this PR

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+      
+      - name: Linters
+        run: docker-compose run test yarn lint
+
+      - name: Unit tests
+        run: docker-compose run test bundle exec rspec --exclude-pattern "spec/systems/**/*_spec.rb" --profile
+
+      - name: System tests
+        run: docker-compose run test bundle exec rspec spec/systems --profile
+
```
Comment (hoangmirs) [inline .github/workflows/test.yml]: It's still missing an end line 😂

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  DOCKER_USERNAME: ${{ github.repository_owner }}
+
+jobs:
+  build:
+    name: Build
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: nimblehq/branch-tag-action@v1
```
Comment (hoangmirs) [inline .github/workflows/test.yml]: You can change to use v1.1 here

File: .github/workflows/test.yml [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+      - name: Docker login to Github package registry
+        run: "echo $DOCKER_TOKEN | docker login $DOCKER_REGISTRY -u $DOCKER_USERNAME --password-stdin"
+
+      - name: Docker pull
+        if: env.BRANCH_TAG != 'master' && env.BRANCH_TAG != 'development'
```
Comment (hoangmirs) [inline .github/workflows/test.yml]: If using branch tag action v1.1, `master` will be changed to `latest`

File: .github/workflows/test.yml [hunk 11]
```diff
@@ -0,0 +1,59 @@
+name: CI
```
Comment (hoangmirs) [inline .github/workflows/test.yml]: Should we name this workflow as `Test` as well?

File: .github/workflows/test.yml [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  build:
+    name: Build
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Branch Tag Action
+        uses: nimblehq/branch-tag-action@v1.1
+
+      - name: Docker login to Github package registry
```
Comment (andyduong1920) [inline .github/workflows/test.yml]: A small thing, is it `Github package registry` 🤔

Comment (gutakk) [reply .github/workflows/test.yml]: It is Github container registry. My bad 😆

Comment (gutakk) [reply .github/workflows/test.yml]: Fixed

File: .github/workflows/test.yml [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    name: Tests
+    runs-on: ubuntu-latest
+    needs: build
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Branch Tag Action
+        uses: nimblehq/branch-tag-action@v1.1
+
+      - name: Docker login to Github container registry
```
Comment (hoangmirs) [inline .github/workflows/test.yml]: As we can set it via the `$DOCKER_REGISTRY` variable, should it be "Docker container registry" instead?

Comment (gutakk) [reply .github/workflows/test.yml]: Make sense, fixed.

**Review by hoangmirs — APPROVED** [review]

**Review by andyduong1920 — APPROVED** [review]
